include/dali/core/small_vector.h

```diff
@@ -0,0 +1,575 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_SMALL_VECTOR_H_
+#define DALI_CORE_SMALL_VECTOR_H_
+
+#include <cuda_runtime.h>
+#include <utility>
+#include "dali/kernels/alloc.h"
+#include "dali/core/cuda_utils.h"
+
+namespace dali {
+
+template <typename T, typename Allocator, bool Contextless = std::is_empty<Allocator>::value>
+struct SmallVectorAlloc {
+  SmallVectorAlloc() = default;
+  __host__ __device__ SmallVectorAlloc(Allocator &&allocator) : allocator_(cuda_move(allocator)) {}
+  __host__ __device__ SmallVectorAlloc(const Allocator &allocator) : allocator_(allocator) {}
+
+  Allocator allocator_;
+  __host__ T *allocate(size_t count) {
+    return allocator_.allocate(count);
+  }
+  __host__ void deallocate(T *ptr, size_t count) {
+    allocator_.deallocate(ptr, count);
+  }
+};
+
+template <typename T, typename Allocator>
+struct SmallVectorAlloc<T, Allocator, true> {
+  static __host__ T *allocate(size_t count) {
+    return Allocator().allocate(count);
+  }
+  static __host__ void deallocate(T *ptr, size_t count) {
+    Allocator().deallocate(ptr, count);
+  }
+};
+
+template <typename T>
+struct SmallVectorAlloc<T, device_side_allocator<T>, true> {
+  static __device__ T *allocate(size_t count) {
+    return device_side_allocator<T>::allocate(count);
+  }
+  static __device__ void deallocate(T *ptr, size_t count) {
+    device_side_allocator<T>::deallocate(ptr, count);
+  }
+};
+
+
+template <typename T, bool is_pod = std::is_pod<T>::value>
+class SmallVectorBase {
+ protected:
+  __host__ __device__ static void move_and_destroy(T *dest, T *src, size_t count) noexcept {
+    for (size_t i = 0; i < count; i++) {
+      new(dest + i) T(cuda_move(src[i]));
+      src[i].~T();
+    }
+  }
+
+  __host__ __device__ static void destroy(T *ptr, size_t count) noexcept {
+    for (size_t i = 0; i < count; i++) {
+      ptr[i].~T();
+    }
+  }
+};
+
+
+template <typename T>
+class SmallVectorBase<T, true> {
+ protected:
+  __host__ __device__ void move_and_destroy(T *dst, T *src, size_t count) noexcept
+  {
+#ifdef __CUDA_ARCH__
+    for (size_t i = 0; i < count; i++) {
+      dst[i] = src[i];
+    }
+#else
+    memcpy(dst, src, count * sizeof(T));
```

Comment:
Isn't there a device side memcpy as well? Is it slow or problematic to dispatch here?

Reply:
It's slow, according to multiple posts on stackoverflow.